Update Environment and post-processing documentation for 4.0 (#6273)
godotengine/godot-docs: tutorials/3d/environment_and_post_processing.rst

Environment and post-processing
===============================

Godot 4 provides a redesigned Environment resource, as well as a new
post-processing system with many available effects right out of the box.

.. note::

    As of Godot 4, Environment *performance/quality* settings are defined in the
    project settings instead of in the Environment resource. This makes global
    adjustments easier, as you no longer have to tweak Environment resources
    individually to suit various hardware configurations.

    Note that most Environment performance/quality settings are only visible
    after enabling the **Advanced** toggle in the Project Settings.

Environment
-----------

The :ref:`class_Environment` resource stores all the information required for

> controlling the 2D and 3D rendering environment. This includes sky, ambient

Reviewer: ```suggestion
controlling the 2D and 3D rendering environment. This includes the sky, ambient
```

    ``default_env.tres`` file that was used in Godot 3 projects.

If no WorldEnvironment node or DirectionalLight3D node is present in the current
scene, the editor will display a preview environment and sun instead. This can
be disabled using the buttons at the top of the 3D editor:

.. image:: img/environment_preview_sun_sky_toggle.webp

Clicking on the 3 vertical dots on the right will display a dialog which allows
you to customize the appearance of the preview environment:

.. image:: img/environment_preview_sun_sky_toggle.webp

**The preview sun and sky is only visible in the editor, not in the running
project.** Using the buttons at the bottom of the dialog, you can add the
preview sun and sky into the scene as nodes.

.. tip::

    If you hold :kbd:`Shift` while clicking **Add Sun to Scene** or **Add
    Environment to Scene** in the preview environment editor, this will add both
    a preview sun and environment to the current scene (as if you clicked both
    buttons separately). Use this to speed up project setup and prototyping.

Camera attributes
-----------------

.. note::

    In Godot 4, exposure and depth of field information was split from the

>     Environment resource into a separate CameraEffects resource. This allows

Reviewer 2: ```suggestion
    Environment resource into a separate CameraAttributes resource. This allows
```

  "hall of mirrors" visual glitch if the sky is visible at any time.

Sky materials
^^^^^^^^^^^^^

When using the **Sky** background mode (or the ambient/reflected light mode is
set to **Sky**), a Sky subresource becomes available to edit in the Environment
resource. Editing this subresource allows you to create a SkyMaterial resource
within the Sky.

There are 3 built-in sky materials to choose from:

- **PanoramaSkyMaterial:** Use a 360 degree panorama sky image (2:1 aspect ratio
  recommended). To benefit from high dynamic range, the panorama image must be
  in an HDR-compatible format such as ``.hdr`` or ``.exr`` rather than a
  standard dynamic range format like ``.png`` or ``.jpg``.
- **ProceduralSkyMaterial:** Use a procedurally generated sky with adjustable
  ground, sun, sky and horizon colors. This is the type of sky used in the
  editor preview. The sun's position is automatically derived from the first 4
  DirectionalLight3D nodes present in the scene. There can be up to 4 suns at a
  given time.
- **PhysicalSkyMaterial:** Use a physically-based procedural sky with adjustable
  scattering parameters. The sun's position is automatically derived from the

>   first DirectionalLight3D nodde present in the scene. PhysicalSkyMaterial is

Reviewer: ```suggestion
  first DirectionalLight3D node present in the scene. PhysicalSkyMaterial is
```

  instead. This improves performance in purely indoor scenes, but creates a
  "hall of mirror" visual glitch if the sky is visible at any time.

Sky materials
^^^^^^^^^^^^^

When using the **Sky** background mode (or the ambient/reflected light mode is
set to **Sky**), a Sky resource becomes available to edit in the Environment
resource. Editing this subresource allows you to create a SkyMaterial resource
within the Sky.

There are 3 built-in sky materials to choose from:

- **PanoramaSkyMateiral:** Use a 360 degree panorama sky image (2:1 aspect ratio
  recommended). To benefit from high dynamic range, the panorama image must be
  in an HDR-compatible format such as ``.hdr`` or ``.exr`` rather than a
  standard dynamic range format like PNG or JPEG.
- **ProceduralSkyMaterial:** Use a procedurally generated sky with adjustable
  ground, sun, sky and horizon colors. This is the type of sky used in the
  editor preview. The sun's position is automatically derived from the
  DirectionalLight3Ds present in the scene.
- **PhysicalSkyMaterial:** Use a physically-based procedural sky with adjustable
  scattering parameters. The sun's position is automatically derived from the

>   first DirectionalLight3Ds present in the scene. PhysicalSkyMaterial is

Reviewer 2: ```suggestion
  first DirectionalLight3D present in the scene. PhysicalSkyMaterial is
```

  slightly more expensive to render compared to ProceduralSkyMaterial.

Panorama sky images are sometimes called HDRIs (High Dynamic Range Images).
You can find freely licensed HDRIs on `Poly Haven <https://polyhaven.com/hdris>`__.

If you need a custom sky material (e.g. for procedural clouds), you can
create a custom :ref:`sky shader <doc_sky_shader>`.

Ambient light
^^^^^^^^^^^^^

Ambient (as defined here) is a type of light that affects every piece of
geometry with the same intensity. It is global and independent of lights that
might be added to the scene. Ambient light is one of the two components of
image-based lighting. Unlike reflected light, ambient light does not vary
depending on the camera's position and viewing angle.

There are several types of ambient light to choose from:

> - **Background:** Source ambient light from the background sky (default).

Reviewer 2: In background mode the source will be whatever the ``Environment.background`` is set to. If the background is set to sky, it will use sky, but if the background is set to clear color it will use the clear color

  Ambient light intensity will vary depending on the sky image's contents, which
  can result in more visually appealing ambient lighting. A sky must be set as
  background for this mode to be visible.
- **Disabled:** Do not use any ambient light. Useful for purely indoor scenes.
- **Color:** Use a constant color for ambient light, ignoring the background
  sky. Ambient light intensity will be the same on all sides, which may result
  in the scene's lighting looking more flat. Useful for indoor scenes where
  pitch black shadows may be too dark, or to maximize performance on low-end
  devices.
- **Sky:** Source ambient light from a specified sky, even if the background is
  set to a mode other than **Sky**. If the background mode is already **Sky**,
  this mode behaves identically to **Background**.

.. image:: img/environment_ambient.png

> When the ambient light mode is set to Background, it's possible to blend between
> the ambient color and sky using the **Sky Contribution** setting. This value is

Reviewer 2: ```suggestion
When the ambient light mode is set to Sky or Background (and background is set
to Sky), it's possible to blend between the ambient color and sky using the
**Sky Contribution** setting. This value is
```

set to ``1.0`` by default, which means that only the ambient sky is used. The
ambient color is ignored unless **Sky Contribution** is decreased below 1.0.

Here is a comparison of how different ambient light affects a scene:

.. image:: img/environment_ambient2.png

Finally, there is an **Energy** setting which is a multiplier. It's useful when
working with HDR.

> In general, ambient light should only be used for simple scenes, large

Reviewer 2: ```suggestion
In general, relying on ambient light alone should only be done for simple scenes, large
```

exteriors, or for performance reasons. Ambient light is fast to render, but it
does not provide the best lighting quality. It's better to generate ambient
light from ReflectionProbe, VoxelGI or SDFGI, as these will simulate how
indirect light propagates more accurately. Below is a comparison, in terms of
quality, between using a flat ambient color and a VoxelGI:

.. image:: img/environment_ambient_comparison.png

Using one of the methods described above, objects get constant ambient lighting
replaced by ambient light from the probes.

Reflected light
^^^^^^^^^^^^^^^

Reflected light (also called specular light) is the other of the two components
of image-based lighting.

Reflected light can be set to one of 3 modes:

> - **Background:** Reflect from the background sky (default).

Reviewer 2: Same as above, Background mode uses whatever the background is set to

- **Disabled:** Do not reflect any light from the environment. Useful for purely
  indoor scenes, or to maximize performance on low-end devices.
- **Sky:** Reflect from the background sky, even if the background is set to a
  mode other than **Sky**. If the background mode is already **Sky**, this mode
  behaves identically to **Background**.

Fog
^^^

.. note::

    This section refers to non-volumetric fog only.
    It is possible to use both non-volumetric fog and :ref:`doc_volumetric_fog`
    at the same time.

> Fog, as in real life, makes distant objects fade away into an uniform color.

Reviewer 2: ```suggestion
Fog, as in real life, makes distant objects fade away into a uniform color.
```

.. image:: img/environment_fog_transmission.png

Volumetric Fog
^^^^^^^^^^^^^^

> Volumetric fog provides a realistic fogging effect to the scene, with fog color

Reviewer 2: ```suggestion
Volumetric fog provides a realistic fog effect to the scene, with fog color
```

  between ``6.0`` and ``8.0``. Higher values result in less blown out highlights,
  but make the scene appear slightly darker as a whole.

Mid- and post-processing effects
--------------------------------

A large amount of widely-used mid- and post-processing effects are supported
in the Environment resource.

.. note::

    Screen-space effects such as SSR, SSAO, SSIL and glow do not operate on
    geometry that is located outside the camera view or is occluded by other
    opaque geometry. Consider this when tweaking their settings to avoid
    distracting changes during gameplay.

Screen-Space Reflections (SSR)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*This feature is only available when using the Clustered Forward backend, not
Forward Mobile or Compatibility.*

While Godot supports three sources of reflection data (Sky, ReflectionProbe, and
VoxelGI), they may not provide enough detail for all situations. Scenarios

> where screen-space reflections make the most sense are when objects are in

Reviewer 2: Godot supports more than 3 sources of reflection data (Sky, ReflectionProbe, VoxelGI, SDFGI, LightmapGI (when using SH)), . I would generalize this to say Godot supports many sources of reflection data, but that SSR provides much higher detail when it works

  between ``6.0`` and ``8.0``. Higher values result in less blown out highlights,
  but make the scene appear slightly darker as a whole.

Mid- and post-processing effects
--------------------------------

The Environment resource supports many widely-used mid- and post-processing effects.

.. note::

    Screen-space effects such as SSR, SSAO, SSIL and glow do not operate on
    geometry that is located outside the camera view or is occluded by other
    opaque geometry. Consider this when tweaking their settings to avoid
    distracting changes during gameplay.

Screen-Space Reflections (SSR)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> *This feature is only available when using the Clustered Forward backend, not
> Forward Mobile or Compatibility.*

Reviewer: ```suggestion
*This feature is only available when using the Forward+ backend, not
Mobile or Compatibility.*
```

This is in other places as well, are those the names of the renderers in the code? I figure we should be using what they're called in editor.

A few user-controlled parameters are available to better tweak the technique:

- **Max Steps:** determines the length of the reflection. The bigger this
  number, the more costly it is to compute.
- **Fade In:** allows adjusting the fade-in curve, which is useful to make the
  contact area softer.
- **Fade Out:** allows adjusting the fade-out curve, so the step limit fades out
  softly.

> - **Depth Tolerance:** can be used for screen-space-ray hit tolerance to gaps.
>   The bigger the value, the more gaps will be ignored.

Reviewer 2: ```suggestion
- **Depth Tolerance:** can be used to allow screen-space-rays to pass behind objects.
  The rays will treat each object as if it has this depth in determining if it can pass behind the object.
```

.. image:: img/environment_ssao.png

It is a common mistake to enable this effect, turn on a light, and not be able to
appreciate it. This is because :abbr:`Screen-Space Ambient Occlusion (SSAO)`
only acts on *ambient* light, not direct light.

This is why, in the image above, the effect is less noticeable under the direct
light (on the left). If you want to force
:abbr:`Screen-Space Ambient Occlusion (SSAO)` to work with direct light too,

> use the **Light Affect** parameter. Even though this is not correct,

Reviewer 2: ```suggestion
use the **Light Affect** parameter. Even though this is not physically correct,
```

  aliasing around the edges of objects. Setting too low will make object edges
  appear blurry.
- **Normal Rejection:** Amount of normal rejection used when calculating
  screen-space indirect lighting. Normal rejection uses the normal of a given
  sample point to reject samples that are facing away from the current pixel.
  Normal rejection is necessary to avoid light leaking when only one side of an
  object is illuminated. However, normal rejection can be disabled if light
  leaking is desirable, such as when the scene mostly contains emissive objects
  that emit light from faces that cannot be seen from the camera.

.. image:: img/environment_ssil.webp

Signed Distance Field Global Illumination (SDFGI)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

*This feature is only available when using the Clustered Forward backend, not
Forward Mobile or Compatibility.*

Signed distance field global illumination (SDFGI) is a form of real-time global
illumination. It is not a screen-space effect, which means it can provide global
illumination for off-screen elements (unlike :abbr:`Screen-Space Indirect Lighting (SSIL)`).

.. seealso::

>     See the SDFGI documentation (TODO) for instructions on setting up this

Reviewer: I assume there should be a link here to the GIProbes page

a given range. It has an initial **Distance** with a **Transition** region
(in world units):

.. image:: img/environment_dof_far.webp

The **Amount** parameter controls the amount of blur. For larger blurs, tweaking
the **Quality** may be needed in order to avoid artifacts.

Depth of Field / Near Blur
^^^^^^^^^^^^^^^^^^^^^^^^^^

This effect simulates focal distance on cameras. It blurs objects close
to the camera (acts in the opposite direction as far blur).
It has an initial **Distance** with a **Transition** region (in world units):

.. image:: img/environment_dof_near.webp

The **Amount** parameter controls the amount of blur. For larger blurs, tweaking
the **Quality** may be needed in order to avoid artifacts.

It is common to use both blurs together to focus the viewer's attention on a
given object, or create a so-called
`"tilt shift" effect <https://en.wikipedia.org/wiki/Miniature_faking>`__.

.. image:: img/environment_mixed_blur.webp

.. note::

    When using CameraAttributesPhysical instead of CameraAttributesPractical,
    depth of field is automatically computed from the camera attributes' focus
    distance, focal length, and aperture.

Exposure
^^^^^^^^

> 

Reviewer: I know there's just one slider but there should be at least a sentence or two here.

The second is by supplying a color correction gradient. This can be done by
assigning a GradientTexture1D resource to the **Color Correction** property, or
by loading a texture containing a horizontal gradient. The leftmost of the
gradient represents black in the source image, whereas the rightmost part of the
gradient represents white in the source image.

A linear black-to-white gradient like the following one will produce no effect:

.. image:: img/environment_adjustments_default_gradient.png

But creating custom ones will allow to map each channel to a different color:

.. image:: img/environment_adjustments_custom_gradient.png

A 3D look-up-texture (LUT) can also be used for color correction. This is a
special texture used to modify each color channel separately from one another
(red, green, blue). This image can be of any resolution, but since color
correction is low-frequency data, sticking to low resolutions is recommended for
performance reasons. A LUT texture's resolution is typically 17×17×17, 33×33×33,
51×51×51 or 65×65×65 (the odd size allows for better interpolation).

.. note::

    Color data within the LUT must be stored in linear space (not

>     perceptual/sRGB), as Godot performs 3D rendering in linear space.

Reviewer 2: This is misleading, color-correction happens in perceptual/sRGB space in Godot. 
https://github.com/godotengine/godot/blob/ea47e03b360d0aeee10c1f7ef798f5750c2b1aa8/servers/rendering/renderer_rd/shaders/effects/tonemap.glsl#L491-L493

3D rendering does happen in linear space but color correction happens after the linear->sRGB conversion in the tonemapping pass

Author: Should I remove the note entirely? I'm not sure how to cause the least confusion here.

Reviewer 2: Yes, might as well remove the note entirely

With the above LUT template, after changing its import mode to **Texture3D**,
set its number of **Horizontal** slices to ``33`` in the Import dock then click
**Reimport**. If you load this LUT into the **Color Correction** property, you
won't see any visible difference for now since this texture is designed to be a
neutral starting point.

This LUT template can be modified in an image editor to provide a different
mood to the image. A common workflow is to place the LUT image next to a
screenshot of the project's 3D viewport, then use an image editor to modify both
the LUT image and the screenshot at the same time. The LUT can then be saved and
applied to the game engine to perform the same color correction in real-time.

For example, with the LUT template modified in an image editor to give it a
"sepia" look, the result on the right is obtained:

.. image:: img/environment_adjustments_3d_lut_comparison.png

.. note::

    Adjustments and color correction are applied *after* tonemapping.
    This means the tonemapping properties defined above still have an effect
    when adjustments are enabled.

Camera attribute options
------------------------

Depth of Field / Far Blur

> ^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 2: There should be a mention here that when using CameraAttributesPhysical, depth of field is calculated automatically from focus distance, focal length, and aperture

This effect simulates focal distance on cameras. It blurs objects behind
a given range. It has an initial **Distance** with a **Transition** region
(in world units):

.. image:: img/environment_dof_far.png

The **Amount** parameter controls the amount of blur. For larger blurs, tweaking
the **Quality** may be needed in order to avoid artifacts.

Depth of Field / Near Blur
^^^^^^^^^^^^^^^^^^^^^^^^^^

This effect simulates focal distance on cameras. It blurs objects close
to the camera (acts in the opposite direction as far blur).
It has an initial **Distance** with a **Transition** region (in world units):

.. image:: img/environment_dof_near.png

The **Amount** parameter controls the amount of blur. For larger blurs, tweaking
the **Quality** may be needed in order to avoid artifacts.

It is common to use both blurs together to focus the viewer's attention on a
given object, or create a so-called
`"tilt shift" effect <https://en.wikipedia.org/wiki/Miniature_faking>`__.

.. image:: img/environment_mixed_blur.png

Exposure
^^^^^^^^

Auto Exposure
^^^^^^^^^^^^^

> *This feature is only available when using the GLES3 backend.*

Reviewer 2: ```suggestion
*This feature is only available when using the Clustered Forward backend, not
Forward Mobile or Compatibility.*
```

different amounts of light.

.. image:: img/environment_hdr_autoexp.gif

The simplest way to use auto exposure is to make sure outdoor lights (or other
strong lights) have energy beyond 1.0. This is done by tweaking their **Energy**
multiplier (on the Light itself). To make it consistent, the **Sky** usually
needs to use the energy multiplier too, to match with the directional light.
Normally, values between 3.0 and 6.0 are enough to simulate indoor-outdoor conditions.

By combining Auto Exposure with :ref:`doc_environment_and_post_processing_glow`
post-processing, pixels that go over the tonemap **White** will bleed to the
glow buffer, creating the typical bloom effect in photography.

.. image:: img/environment_hdr_bloom.png

The user-controllable values in the Auto Exposure section come with sensible
defaults, but you can still tweak them:

.. image:: img/environment_hdr.png

- **Scale:** Value to scale the lighting. Brighter values produce brighter
  images, smaller ones produce darker ones.
- **Min Luma:** Minimum luminance that auto exposure will aim to adjust for.
  Luminance is the average of the light in all the pixels of the screen.

> - **Max Luma:** Maximum luminance that auto exposure will aim to adjust for.

Reviewer 2: These should be renamed given their respective names in CameraAttributesPhysical and CameraAttributesPractical (max_exposure_value/max_sensitivity)
